stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,1455 @@
+                                    );
+                                    transaction
+                                }
+                                Err(e) => {
+                                    warn!("Signer #{}: Failed to cast aggregate public key vote: {:?}", self.signer_id, e);
+                                    continue;
+                                }
+                            }
+                        }
+                        EpochId::Epoch30 => {
+                            debug!("Signer #{}: Received a DKG result, but are in epoch 3. Broadcast the transaction to stackerDB.", self.signer_id);
+                            match retry_with_exponential_backoff(|| {
+                                self.stacks_client
+                                    .build_vote_for_aggregate_public_key(
+                                        self.reward_cycle,
+                                        self.stackerdb.get_signer_slot_id(),
+                                        point.clone(),
+                                    )
+                                    .map_err(backoff::Error::transient)
+                            }) {
+                                Ok(transaction) => transaction,
+                                Err(e) => {
+                                    warn!("Signer #{}: Failed to build a cast aggregate public key vote transaction: {:?}", self.signer_id, e);
+                                    continue;
+                                }
+                            }
+                        }
+                    };
+                    let old_transactions = self
+                        .stackerdb
+                        .get_signer_transactions_with_retry(&vec![self.signer_id])
+                        .map_err(|e| {
+                            error!("Failed to get old transactions from stackerdb: {:?}", e);
+                        })
+                        .unwrap_or_default();
+                    // Filter out our old transactions that are no longer valid
+                    let mut new_transactions: Vec<_> = old_transactions.into_iter().filter_map(|transaction|  {
+                        let origin_address = transaction.origin_address();
+                        let origin_nonce = transaction.get_origin_nonce();
+                        let Ok(account_nonce) = retry_with_exponential_backoff(|| self.stacks_client.get_account_nonce(&origin_address).map_err(backoff::Error::transient)) else {
+                            warn!("Signer #{}: Unable to get account for address: {origin_address}. Removing {} from our stored transactions.", self.signer_id, transaction.txid());
+                            return None;
+                        };
+                        if origin_nonce < account_nonce {
+                            debug!("Signer #{}: Transaction {} has an invalid nonce. Removing if removing it from our stored transactions.", self.signer_id, transaction.txid());
```

Comment:
```suggestion
                            debug!("Signer #{}: Transaction {} has an invalid nonce. Removing it from our stored transactions.", self.signer_id, transaction.txid());
```

Reply:
Also, am I correct that the logic here is more like "this transaction was already confirmed so we can't update it"?

Reply:
It is both. it could be invalid or it was already confirmed. We do not know which without checking the mempool.

Reply:
I will change it to "has outdated nonce"